src/libs/kata-sys-util/src/fs.rs

```diff
@@ -0,0 +1,234 @@
+    let src_path = src.as_ref();
+    let dst_path = dst.as_ref();
+    let src = src_path.to_string_lossy();
+    let dst = dst_path.to_string_lossy();
+
+    if !src_path.is_file() {
+        return Err(eother!("reflink_copy src {} is not a regular file", src));
+    }
+
+    // Make sure dst's parent exist. If dst is a regular file, then unlink it for later copy.
+    if dst_path.exists() {
+        if !dst_path.is_file() {
+            return Err(eother!("reflink_copy dst {} is not a regular file", dst));
+        } else {
+            fs::remove_file(dst_path)?;
+        }
+    } else if let Some(dst_parent) = dst_path.parent() {
+        if !dst_parent.exists() {
+            if let Err(e) = fs::create_dir_all(dst_parent) {
+                return Err(eother!(
+                    "reflink_copy: create_dir_all {} failed: {:?}",
+                    dst_parent.to_str().unwrap(),
+                    e
+                ));
+            }
+        } else if !dst_parent.is_dir() {
+            return Err(eother!("reflink_copy parent of {} is not a directory", dst));
+        }
+    }
+
+    // Reflink copy, and fallback to regular copy if reflink fails.
+    let src_file = fs::File::open(src_path)?;
+    let dst_file = fs::File::create(dst_path)?;
+    if let Err(e) = do_reflink_copy(src_file, dst_file) {
+        match e.raw_os_error() {
+            // Cross dev copy or filesystem doesn't support reflink, do regular copy
+            Some(os_err)
+                if os_err == nix::Error::EXDEV as i32
+                    || os_err == nix::Error::EOPNOTSUPP as i32 =>
+            {
+                warn!(
+                    sl!(),
+                    "reflink_copy: reflink is not supported ({:?}), do regular copy instead", e,
+                );
+                if let Err(e) = do_regular_copy(src.as_ref(), dst.as_ref()) {
+                    return Err(eother!(
+                        "reflink_copy: regular copy {} to {} failed: {:?}",
+                        src,
+                        dst,
+                        e
+                    ));
+                }
+            }
+            // Reflink copy failed
+            _ => {
+                return Err(eother!(
+                    "reflink_copy: copy {} to {} failed: {:?}",
+                    src,
+                    dst,
+                    e,
+                ))
+            }
+        }
+    }
+
+    Ok(())
+}
+
+// Copy file using cp command, which handles sparse file copy.
+fn do_regular_copy(src: &str, dst: &str) -> Result<()> {
+    let mut cmd = Command::new("/bin/cp");
+    cmd.args(&["--sparse=auto", src, dst]);
+
+    match cmd.output() {
+        Ok(output) => match output.status.success() {
+            true => Ok(()),
+            false => Err(eother!("`{:?}` failed: {:?}", cmd, output)),
+        },
+        Err(e) => Err(eother!("`{:?}` failed: {:?}", cmd, e)),
+    }
+}
+
+/// Copy file by reflink
+fn do_reflink_copy(src: File, dst: File) -> Result<()> {
```

Comment:
- What if `src` == `dst`?
- What if `src is actually a link?

Reply:
These cases are protected at the entrance of `reflink_copy()`.